pkg/frontend/querymiddleware/remote_read.go

```diff
@@ -229,14 +291,49 @@ func (r *remoteReadQueryRequest) WithQuery(_ string) (MetricsQueryRequest, error
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithHeaders([]*PrometheusHeader) MetricsQueryRequest {
+func (r *remoteReadQueryRequest) WithHeaders(_ []*PrometheusHeader) MetricsQueryRequest {
 	panic("not implemented")
 }
 
-func (r *remoteReadQueryRequest) WithStartEnd(_ int64, _ int64) (MetricsQueryRequest, error) {
-	panic("not implemented")
+// WithStartEnd clones the current remoteReadQueryRequest with a new start and end timestamp.
+func (r *remoteReadQueryRequest) WithStartEnd(start int64, end int64) (MetricsQueryRequest, error) {
+	clonedQuery, err := cloneRemoteReadQuery(r.query)
+	if err != nil {
+		return nil, err
+	}
+
+	clonedQuery.StartTimestampMs = start
+	clonedQuery.EndTimestampMs = end
+
+	// We only clamp the hints time range (and not extend it). If, for any reason, the hints start/end
+	// time range is shorter than the query start/end range, then we manipulate only to clamp it to keep
+	// it within the requested range.
+	if clonedQuery.Hints != nil && clonedQuery.Hints.StartMs < start {
```

Comment:
nit: now I assume that if someone provided hints, they set a sensible  `Hints.StartMs`. Since we don't have validation midddleware, maybe we do need a zero check
```suggestion
	if clonedQuery.Hints != nil && clonedQuery.Hints.StartMs != 0 && clonedQuery.Hints.StartMs < start {
```
Just to avoid setting start > end

Reply:
I'm not super sure about this.

_Let's forget for a moment the fact that today we're ignoring `Hints` in Mimir, and let's assume we'll fix it (as I think we should do)._

The reason why I think we should not consider `0` an "invalid" number is because, if hints are populated (non nil), we should just honor values are they're set. Prometheus does the same in its remote read implementation, where `ReadHints` are converted into `SelectHints`  and then start/end from `SelectHints` are just used as is (and we do the same in Mimir for instant/range queries):
https://github.com/prometheus/prometheus/blob/5efc8dd27b6e68d5102b77bc708e52c9821c5101/tsdb/querier.go#L135-L136